### ISSUE TITLE
Support for ParamSpec and overloaded signatures

### ISSUE THREAD

**Issue by tjsmart** [microsoft/pyright#4572]
First thank you so much pyright team for all that you do for the python community!! It's greatly appreciated!

**Is your feature request related to a problem? Please describe.**
I am trying to apply ParamSpec to an overloaded function such as the simplified example below:

**Example setup**
Setup:
```bash
$ python --version
Python 3.8.10
$ pyright --version
pyright 1.1.292
$ python -m pip list | grep typing-extensions
typing-extensions         4.4.0
```

Example file:
```python
from __future__ import annotations

from typing import Callable
from typing import overload
from typing import TypeVar

from typing_extensions import ParamSpec

P = ParamSpec("P")
R = TypeVar("R")


def simple_callable(
    func: Callable[P, R], *args: P.args, **kwargs: P.kwargs
) -> Callable[[], R]:
    ...


@overload
def foo() -> None:
    ...


@overload
def foo(a: int) -> None:
    ...


def foo(a: int = 1) -> None:
    ...
```

**Describe the solution you'd like**

I'd hope for pyright to have the below behavior:

```python
simple_callable(foo)         # expected this to pass, overload without arguments exists
simple_callable(foo, 1)      # expected this to pass, overload with a single positional argument exists
simple_callable(foo, a=1)    # expected this to pass, overload with keyword argument 'a' exists
simple_callable(foo, 1, 2)   # expected this to fail due to no overload with two positional arguments existing
simple_callable(foo, b=2)    # expected this to fail due to no overload with the keyword argument 'b' existing
```

**Additional context**

Current behavior that I observe is an error message such as below:
```
error: Argument of type "Overload[() -> None, (a: int) -> None, (a: int = 1) -> None]" cannot be assigned to parameter "func" of type "(**P@simple_callable) -> R@simple_callable" in function "simple_callable"
    ParamSpec cannot be used with overloaded function (reportGeneralTypeIssues)
error: Param spec "P@simple_callable" has no bound value (reportGeneralTypeIssues)
```

Also please note that the above example is overly simplified. I recognize that the overloads of foo do not add much if any value.

**Comment by erictraut** [comment]
`ParamSpec` cannot be used with overloaded functions. PEP 612, which introduced `ParamSpec` unfortunately didn't contemplate this combination, and it's unclear how they would work together. A `ParamSpec` must be matched against a single function signature, but an overload is a set of two or more distinct signatures. 

Mypy's implementation of `ParamSpec` allows the use of an overloaded function, but it captures only the first overloaded signature. This behavior seems rather random to me — and is probably not consistent with the intent of the coder, so I chose to make this condition an error in pyright.

**Comment by tjsmart** [comment]
Thanks for the reply @erictraut, but I'm not sure I really understand. Shouldn't ParamSpec just use the first overload that matches successfully? And if no overload matches successfully, report an error?


Essentially in my example above, the behavior I expected isn't too different from just using the overloaded function directly:
```python
foo()      # no error
foo(1)     # no error
foo(a=1)   # no error
foo(1, 2)  # error: No overloads for "foo" match the provided arguments    Argument types: (Literal[1], Literal[2])
foo(b=2)   # error: No overloads for "foo" match the provided arguments    Argument types: (Literal[2])
```

Hopefully sometime this can be improved as many builtin functions are overloaded which makes them unusable with ParamSpec + pyright.

**Comment by erictraut** [comment]
> Shouldn't ParamSpec just use the first overload that matches successfully?

I don't know what you mean by that.

The issue is with `simple_callable`, which is a generic function. If you pass `foo` as an argument to `simple_callable`, the constraint solver needs to solve the type parameters `P` and `R`. But if `foo` has multiple overloaded signatures, there is no solution for `P` and `R`.

**Comment by tjsmart** [comment]
In the below example, since the `Literal[1]` overload appears first `bar(1)` will be of type `str`.
```python
@overload
def bar(x: Literal[1]) -> str: ...

@overload
def bar(x: int) -> list: ...

def bar(x: int) -> list | str:
    ...

reveal_type(bar(1))   #  Type of "bar(1)" is "str"
```

However if I reverse the order of the overloaded signatures I will get the opposite result:
```python
@overload
def bar(x: int) -> list: ...

@overload
def bar(x: Literal[1]) -> str: ...

def bar(x: int) -> list | str:
    ...

reveal_type(bar(1))   #  Type of "bar(1)" is "list[Unkown]"
```

For `ParamSpec` I imagine in many cases (since it is being generic) it will use the first available call signature, but if there is a type error like in my original example, it should start checking other call signatures to find an overload that applies.

> The issue is with simple_callable, which is a generic function. If you pass foo as an argument to simple_callable, the constraint solver needs to solve the type parameters P and R. But if foo has multiple overloaded signatures, there is no solution for P and R.

Not sure why there needs to be a single solution for `P` and `R`. Doesn't seem to match how overloads typically work in python.

**Comment by erictraut** [comment]
As I mentioned, PEP 612 doesn't appear to contemplate the use of overloaded functions with ParamSpecs. It doesn't describe how this would work. I've thought about the problem more, and I found a way to accommodate overloads with ParamSpecs in a limited manner. In particular, in cases where a ParamSpec is scoped to a function and is used in a `Callable` within the return type annotation or the function uses `P.args` and `P.kwargs` parameters, as in your sample above.

I've augmented pyright's constraint solver so it is now capable of solving distinct sets of TypeVars and ParamSpecs, one for each combination of overloaded signatures. (Note that I said "combination" because multiple ParamSpecs could be used in the same function, and each could be associated with an overloaded function. In this case, all combinatorics of the multiple overloads would need to be considered.)

I don't see a way to make this work in cases where a ParamSpec is scoped to a generic class or generic type alias. There is no mechanism in the Python type system to describe an "overloaded" class or type alias. The concept of an overload applies only to functions.

**Comment by tjsmart** [comment]
Wow, thanks Eric! What you said makes sense to me, although I haven't thought much about (or had a use case) to apply `ParamSpec` to a generic class so I'll take your word for it.

Can't wait to try out the next version!

**Comment by tjsmart** [comment]
> I don't see a way to make this work in cases where a ParamSpec is scoped to a generic class or generic type alias. There is no mechanism in the Python type system to describe an "overloaded" class or type alias. The concept of an overload applies only to functions.

Still trying to understand this point. If I were to rewrite the original `ParamSpec` function `simple_callable` as a generic class, such as `SimpleCallable` below, applying this to overloaded functions will not be supported?
```python
class SimpleCallable(Generic[P, R]):
    def __init__(
        self, fn: Callable[P, R], *args: P.args, **kwargs: P.kwargs
    ) -> None:
        self.fn = fn
        self.args = args
        self.kwargs = kwargs

    def __call__(self) -> R:
        return self.fn(*self.args, **self.kwargs)


SimpleCallable(foo)   # this will give an error of not supported even in the next version?
```

**Comment by erictraut** [comment]
This example will work fine (no errors produced, and behavior as you would expect) because the `*args` and `**kwargs` parameters eliminate all but a single overload.

There are other cases that won't work as well — in particular, cases where multiple overloads match a class-bound ParamSpec. Consider the case where we modify `SimpleCallable` to remove the `*args` and `**kwargs` parameters from the constructor and move them to the `__call__` method. 

```python
class SimpleCallable(Generic[P, R]):
    def __init__(self, fn: Callable[P, R]) -> None:
        self.fn = fn

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R:
        return self.fn(*args, **kwargs)


fn = SimpleCallable(foo)
reveal_type(fn) # SimpleCallable[[], None]
fn() # Works
fn(1) # Error
```

In this case, the specialized instance of `SimpleCallable` assigned to `fn` is `SimpleCallable[[], None]`. In other words, its specialized type considers only the first of the two matching overloads. I think that's reasonable behavior given the limitations of the type system.

**Comment by tjsmart** [comment]
Ahh that makes sense. Thanks Eric for explaining that last point!

**Comment by erictraut** [comment]
This is included in pyright 1.1.293, which I just published. It will also be included in a future release of pylance.
